### PR TITLE
Implement my shot introduction view. Closes #1968

### LINKED ISSUE #1968: Experiment: Screenshot on-boarding (Drive users towards MyShots)
Idea from Max; show a blue dot when users has unread screenshots in MyShots.

### PR DESCRIPTION
**Pull request by josephyu1110**

### CODE REVIEW

**Comment by josephyu1110** [comment]
- For the updated Rocket users, see them as first time using screenshot feature, so show onboarding and the menu after they click screenshot button
- No show “screenshot saved” toast for the first time screenshot
- When opened from external app, hide onboarding and the menu, and open a new tab

File: app/src/main/java/org/mozilla/focus/utils/DialogUtils.java [hunk 1]
```diff
@@ -200,4 +204,50 @@ public static void showPrivacyPolicyUpdateNotification(Context context) {
...
+        params.height = targetView.getMeasuredHeight();
+        params.setMargins(location[0], location[1] - ViewUtils.getStatusBarHeight(activity), 0, 0);
+
+        builder.setView(root);
+
+        final Dialog dialog = builder.create();
+        dialog.show();
+
+        // Press back key will dismiss on boarding view and menu view
+        dialog.setOnCancelListener(dialog1 -> ((MainActivity) activity).dismissAllMenus());
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/utils/DialogUtils.java]: Casting to MainActivity is error prone and may change in the future.

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.java [hunk 2]
```diff
@@ -1033,4 +1041,10 @@ public TabView create() {
             return (TabView) WebViewProvider.create(this.activity, null);
         }
     }
+
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/activity/MainActivity.java]: Add @UiThread

File: app/src/main/java/org/mozilla/focus/utils/ViewUtils.java [hunk 3]
```diff
@@ -76,4 +79,11 @@ public static void updateAlphaIfViewExists(@Nullable Activity activity, @IdRes i
 
         view.setAlpha(alpha);
     }
+
+    public static int getStatusBarHeight(@NonNull final Activity activity) {
+        Rect rectangle = new Rect();
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/utils/ViewUtils.java]: nit: final

File: app/src/main/java/org/mozilla/focus/utils/ViewUtils.java [hunk 4]
```diff
@@ -76,4 +79,11 @@ public static void updateAlphaIfViewExists(@Nullable Activity activity, @IdRes i
...
+        Window window = activity.getWindow();
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/utils/ViewUtils.java]: ditto

File: app/src/main/java/org/mozilla/focus/widget/FocusView.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import android.graphics.Path
+import android.support.v4.content.ContextCompat
+import android.util.AttributeSet
+import android.view.View
+
+import org.mozilla.focus.R
+import org.mozilla.focus.utils.ViewUtils
+
+class FocusView : View {
+    private var transparentPaint: Paint? = null
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/widget/FocusView.kt]: nit: lateinit

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 6]
```diff
@@ -1363,4 +1365,16 @@ private String getLastInsertedUrl(Tab tab) {
...
         }
     }
+
+    public void onCaptureComplete() {
+        final Activity activity = getActivity();
+        if (activity == null) {
+            return;
+        }
+        if (!Settings.getInstance(activity).getEventHistory().contains(Settings.Event.ShowMyShotOnBoardingDialog)) {
+            Settings.EventHistory history = Settings.getInstance(activity).getEventHistory();
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: redundant Settings.getInstance(activity).getEventHistory().

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 7]
```diff
@@ -1363,4 +1365,16 @@ private String getLastInsertedUrl(Tab tab) {
...
+            history.add(Settings.Event.ShowMyShotOnBoardingDialog);
+            ((MainActivity) activity).showMyShotOnBoarding();
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: Casting to MainActivity is dangerous. Is there a better way?

File: app/src/main/java/org/mozilla/focus/utils/DialogUtils.java [hunk 8]
```diff
@@ -200,4 +204,50 @@ public static void showPrivacyPolicyUpdateNotification(Context context) {
...
+        int [] location = new int[2];
+        int [] center = new int[2];
+        // Get target view's position
+        targetView.getLocationInWindow(location);
+
+        // Get spotlight circle's center
+        center[0] = location[0] + targetView.getMeasuredWidth() / 2;
+        center[1] = location[1] + targetView.getMeasuredHeight() / 2;
+
+        AlertDialog.Builder builder = new AlertDialog.Builder(activity, R.style.TabTrayTheme);
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/utils/DialogUtils.java]: nit: final

File: app/src/main/java/org/mozilla/focus/utils/DialogUtils.java [hunk 9]
```diff
@@ -200,4 +204,50 @@ public static void showPrivacyPolicyUpdateNotification(Context context) {
...
+        // Inflate my shot on boarding view
+        ViewGroup root = (ViewGroup) LayoutInflater.from(activity).inflate(R.layout.myshot_onboarding, null);
+
+        // Initialize FocusView and add it to my shot on boarding view's index 0(the bottom of Z-order)
+        FocusView focusView = new FocusView(activity, center, activity.getResources().getDimensionPixelSize(R.dimen.myshot_focus_view_radius));
+        root.addView(focusView, 0);
+
+        // Add a mock my shot menu view to determine the position of hint image and text. Also consuming the click event.
+        View myShotMockView = root.findViewById(R.id.my_shot_mock_menu);
+        RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) myShotMockView.getLayoutParams();
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/utils/DialogUtils.java]: nit: final

File: app/src/main/java/org/mozilla/focus/utils/DialogUtils.java [hunk 10]
```diff
@@ -200,4 +204,50 @@ public static void showPrivacyPolicyUpdateNotification(Context context) {
...
+        targetView.getLocationInWindow(location);
+
+        // Get spotlight circle's center
+        center[0] = location[0] + targetView.getMeasuredWidth() / 2;
+        center[1] = location[1] + targetView.getMeasuredHeight() / 2;
+
+        AlertDialog.Builder builder = new AlertDialog.Builder(activity, R.style.TabTrayTheme);
+
+        // Inflate my shot on boarding view
+        ViewGroup root = (ViewGroup) LayoutInflater.from(activity).inflate(R.layout.myshot_onboarding, null);
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/utils/DialogUtils.java]: nit: final

File: app/src/main/java/org/mozilla/focus/utils/DialogUtils.java [hunk 11]
```diff
@@ -200,4 +204,50 @@ public static void showPrivacyPolicyUpdateNotification(Context context) {
...
+
+        // Initialize FocusView and add it to my shot on boarding view's index 0(the bottom of Z-order)
+        FocusView focusView = new FocusView(activity, center, activity.getResources().getDimensionPixelSize(R.dimen.myshot_focus_view_radius));
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/utils/DialogUtils.java]: nit: final

File: app/src/main/java/org/mozilla/focus/widget/FocusView.kt [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    private val path = Path()
+    private var center: IntArray? = null
```
Comment (cnevinc) [inline app/src/main/java/org/mozilla/focus/widget/FocusView.kt]: Can it be non-null?

**Pull request merged**